File: Problem-1/Problem1.js
// Advent of Code 2020 Day 1

const fs = require('fs');
const data = fs.readFileSync('./Problem-1.txt').toString().split('\n');
const arr = data.map((val) => Number(val));

const solution = (array, target) => {
    let left = 0, right = array.length - 1;
    while (left < right){
        if (array[left] + array[right] === target){
            return [left, right];
        }
        else if (array[left] + array[right] < target){
            left++;
        }
        else{
            right--;
        }
    }
}

const part1 = (arr) => {
    return solution(arr.sort(), 2020).reduce((acc, index) => {
        return acc * arr[index];
    }, 1)
};

console.log(part1(arr));

const hash_map = {}
arr.forEach((val, index)=>{
    hash_map[val] = index;
});
const partTwo = (array, target) => {
    for(i in hash_map){
        let b = target - i;
        const res = solution(array, target - i);
        if (res !== undefined){
            if(res[0] !== array[i] && res[1] !== array[i]){
                return i * array[res[0]] * array[res[1]];
            }
        }
    }
}
console.log(partTwo(arr.sort(), 2020));

File: Problem-1/Problem1.py
#  Advent of Code 2020 Day 1

with open('./Problem-1.txt') as file:
    data = file.read().split('\n')
    arr = sorted([int(ele) for ele in data])

# Solution for part 1
def partOne(arr, target):
    l, r = 0, len(arr) - 1
    while l < r:
        if arr[l] + arr[r] == target:
            return l, r
        elif arr[l] + arr[r] < target:
            l += 1
        else:
            r -= 1

sol = partOne(arr, 2020)
print(arr[sol[0]] * arr[sol[1]])

# Solution for part 2
def partTwo(arr, target):
    d = {}
    for index, val in enumerate(arr, 0):
        d[val] = index
    for key, value in d.items():
        rem = target - key
        sol = partOne(arr, rem)
        if sol:
            if sol[0] != val and sol[1] != val:
                return key * arr[sol[0]] * arr[sol[1]]


print(partTwo(arr, 2020))

